multi/stm32l4-multi/libmulti/libdma.c

```diff
@@ -81,110 +164,145 @@ static void _configureChannel(int dma, int channel, int dir, int priority, void
 		irqnum = 16 + dmainfo[dma].irq_base + channel;
 	}
 
-	interrupt(irqnum, irqHandler, (void *)dma, dma_common[dma].cond, NULL);
+	interrupt(irqnum, irqHandler, (void *)dma, interruptCond, NULL);
 
 	*(channel_base + ccr) = ((priority & 0x3) << 12) | ((msize & 0x3) << 10) | ((psize & 0x3) << 8) |
 		((minc & 0x1) << 7) | ((pinc & 0x1) << 6) | ((dir & 0x1) << 4);
-	*(channel_base + cpar) = (unsigned int) paddr;
+	*(channel_base + cpar) = (unsigned int)paddr;
 	tmp = *(dma_common[dma].base + cselr) & ~(0xF << channel * 4);
-	*(dma_common[dma].base + cselr) = tmp | ((unsigned int) reqmap << channel * 4);
+	*(dma_common[dma].base + cselr) = tmp | ((unsigned int)reqmap << channel * 4);
 }
 
 
-int libdma_configureSpi(int num, int dir, int priority, void *paddr, int msize, int psize, int minc, int pinc)
+int libdma_configurePeripheral(const struct libdma_per *per, int dir, int priority, void *paddr, int msize, int psize, int minc, int pinc, handle_t *cond)
 {
-	int dma = spiChanMap[num].dma;
-	int channel = spiChanMap[num].channel[dir];
-	char reqmap = spiChanMap[num].reqmap;
+	int dma = per->dma, channel = per->channel[dir];
+	char reqmap = per->reqmap;
 
-	_configureChannel(dma, channel, dir, priority, paddr, msize, psize, minc, pinc, reqmap);
+	_configureChannel(dma, channel, dir, priority, paddr, msize, psize, minc, pinc, reqmap, cond);
 
 	return 0;
 }
 
 
-static void _prepare_transfer(volatile unsigned int *channel_base, void *maddr, size_t len, int tcie)
-{
-	*(channel_base + cmar) = (unsigned int) maddr;
-	*(channel_base + cndtr) = len;
-	dataBarier();
-
-	if (tcie)
-		/* Enable Transfer Complete interrupt, enable channel */
-		*(channel_base + ccr) |= (1 << 1) | 0x1;
-	else
-		/* Enable channel */
-		*(channel_base + ccr) |= 0x1;
-
-	dataBarier();
-}
-
-
-static void _unprepare_transfer(volatile unsigned int *channel_base)
-{
-	dataBarier();
-	/* Disable Transfer Complete interrupt, disable channel */
-	*(channel_base + ccr) &= ~((1 << 1) | 0x1);
-	dataBarier();
-}
-
-
 static int _has_channel_finished(const void *maddr, volatile unsigned int *channel_base)
 {
-	return (maddr != NULL && *(channel_base + cndtr) > 0);
+	return (maddr != NULL) && (*(channel_base + cndtr) > 0);
 }
 
 
 static int _transfer(int dma, int rx_channel, int tx_channel, void *rx_maddr, const void *tx_maddr, size_t len)
 {
 	/* Empirically chosen value to avoid mutex+cond overhead for short transactions */
 	int use_interrupts = len > 24;
+	int interrupts_flags = (use_interrupts == 0) ? 0 : DMA_TCIE_FLAG;
+
+	volatile unsigned int *rx_channel_base = dma_common[dma].base + 2 + (5 * rx_channel);
+	volatile unsigned int *tx_channel_base = dma_common[dma].base + 2 + (5 * tx_channel);
 
-	volatile unsigned int *rx_channel_base = dma_common[dma].base + 2 + 5 * rx_channel;
-	volatile unsigned int *tx_channel_base = dma_common[dma].base + 2 + 5 * tx_channel;
+	assert(DMA_MAX_LEN >= len);
```

Comment:
I'd rather not have `multidrv` randomly shutdown because of data passed by a user

Reply:
Applies to all asserts in `multidrv`